lib/propolis/src/hw/ps2ctrl.rs

```diff
@@ -10,73 +11,241 @@ use crate::migrate::*;
+
+#[usdt::provider(provider = "propolis")]
+mod probes {
+    fn ps2ctrl_keyevent(
+        keysym_val: u32,
+        scan_code_set: u8,
+        s0: u8,
+        s1: u8,
+        s2: u8,
+        s3: u8,
+    ) {
+    }
+
+    fn ps2ctrl_keyevent_dropped(
+        keysym_val: u32,
+        is_pressed: u8,
+        scan_code_set: u8,
+    ) {
+    }
+}
 
 bitflags! {
+    /// Controller Status Register
+    ///
+    /// An 8-bit register indicating the status of the controller, accessed by reading from the
+    /// Control Port.
     #[derive(Default)]
     pub struct CtrlStatus: u8 {
+        /// Output Buffer Status
+        /// This bit must be set to 1 (indicating the buffer is full) before the OS attempts to
+        /// read data from the data port.
         const OUT_FULL = 1 << 0;
+
+        /// Input Buffer Status
+        /// 0 if the input buffer is empty; 1 if the input buffer is full and shouldn't be written
+        ///   to by the OS.
+        /// XXX(JPH): Not sure if this should be used somewhere.
         const IN_FULL = 1 << 1;
+
+        /// System Flag
+        /// This bit should be cleared to 0 by the controller on reset, and set to 1 if the system
+        /// passes self tests.
         const SYS_FLAG = 1 << 2;
+
+        /// Command/Data
+        /// 1 if the last write to the input buffer (data port) was a command; 0 if the last write to the input
+        ///   buffer was data.
         const CMD_DATA = 1 << 3;
+
+        /// Keyboard Locked
         const UNLOCKED = 1 << 4;
+
+        /// Auxiliary Output Buffer contains data
         const AUX_FULL = 1 << 5;
+
+        /// Timeout Error (0 = no error, 1 = timeout error)
         const TMO = 1 << 6;
+
+        /// Parity Error with last byte (0 = no error, 1 = timeout error)
         const PARITY = 1 << 7;
     }
 }
 
 bitflags! {
+    /// Controller Configuration Byte
+    /// The OS can read and write this byte to configure the controller.
     #[derive(Default)]
     pub struct CtrlCfg: u8 {
+        /// Primary Port Interrupt (1 = enabled, 0 = disabled)
         const PRI_INTR_EN = 1 << 0;
+
+        /// Auxiliary Port Interrupt (1 = enabled, 0 = disabled)
         const AUX_INTR_EN = 1 << 1;
+
+        /// System Flag (1 = system tests passed)
         const SYS_FLAG = 1 << 2;
+
+        // bit 3: must be 0
+
+        /// Primary Port Clock (1 = disabled, 0 = enabled)
         const PRI_CLOCK_DIS = 1 << 4;
+
+        /// Auxiliary Port Clock (1 = disabled, 0 = enabled)
         const AUX_CLOCK_DIS = 1 << 5;
+
+        /// Primary Port Translation (1 = enabled, 0 = disabled)
+        /// If enabled, the controller should translate keyboard data to scan code set 1.
         const PRI_XLATE_EN = 1 << 6;
+
+        // bit 7: must be 0
     }
 }
 
 bitflags! {
+    /// Controller Output Port
     #[derive(Default)]
     pub struct CtrlOutPort: u8 {
+        // bit 0: system reset
+        // should always be set to 1
+        // XXX(JPH): why does this work
```

Comment:
We are doing none of the work to emulate A20.  Its function is a big lie if a guest actually tries to use it.